rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,592 @@
+        # with a replay buffer. This is besides its existence inside
+        # the RolloutWorkers for training and evaluation.
+        "max_num_policies_to_train": 11,
+
+        # Basic "multiagent" setup for league-building AlphaStar:
+        # Start with "main_0" (the policy we would like to use in the very end),
+        # "league_exploiter_0" (a random policy) and "main_exploiter_0"
+        # (also random).
+        # Once "main_0" reaches n% win-rate, we'll create a snapshot of it
+        # (main_1) as well as create learning "league_exploiter_1" and
+        # "main_exploiter_1" policies (cloned off "main_0").
+        "multiagent": {
+            # Initial policy map. This will be expanded
+            # to more policy snapshots inside the `build_league` method.
+            "policies": {
+                # Our main policy, we'd like to optimize.
+                "main_0": PolicySpec(),
+                # Initial main exploiters (1 of these is a random policy).
+                "main_exploiter_0": PolicySpec(policy_class=RandomPolicy),
+                "main_exploiter_1": PolicySpec(),
+                "main_exploiter_2": PolicySpec(),
+                "main_exploiter_3": PolicySpec(),
+                "main_exploiter_4": PolicySpec(),
+                # Initial league exploiter (1 of these is a random policy).
+                "league_exploiter_0": PolicySpec(policy_class=RandomPolicy),
+                "league_exploiter_1": PolicySpec(),
+                "league_exploiter_2": PolicySpec(),
+                "league_exploiter_3": PolicySpec(),
+                "league_exploiter_4": PolicySpec(),
+            },
+            "policy_mapping_fn":
+                (lambda aid, ep, worker, **kw: "main_0" if
+                 ep.episode_id % 2 == aid else "main_exploiter_0"),
+            # Train all non-random policies that exist at beginning.
+            "policies_to_train": [
+                "main_0", "main_exploiter_1", "main_exploiter_2",
+                "main_exploiter_3", "main_exploiter_4", "league_exploiter_1",
+                "league_exploiter_2", "league_exploiter_3",
+                "league_exploiter_4",
+            ],
+        },
+
+        # Reporting interval.
+        "min_time_s_per_reporting": 2,
+
+        # Use the `training_iteration` method instead of an execution plan.
+        "_disable_execution_plan_api": True,
+    },
+    _allow_unknown_configs=True,
+)
+
+# __sphinx_doc_end__
+# yapf: enable
+
+
+class AlphaStarTrainer(appo.APPOTrainer):
+    @classmethod
+    @override(Trainer)
+    def default_resource_request(cls, config):
+        cf = dict(cls.get_default_config(), **config)
+
+        num_policies = len(cf["max_num_policies_to_train"])
+        if cf["num_gpus"]:
+            num_learner_shards = min(cf["num_gpus"], num_policies)
+            num_gpus_per_shard = cf["num_gpus"] / num_learner_shards
+        else:
+            num_learner_shards = cf.get("num_replay_buffer_shards", 1)
```

Comment:
may be a nit.
can we rename num_replay_buffer_shards to num_learners?
we construct a rb for each leaner, not the other way around. mentally, it feels easier when rb serves shards, and not shards serves rb ...

Reply:
Yeah, wasn't sure about the naming here either. A "shard" is a unit with:
- n GPUs (all on same machine)
- 1 replay buffer actor (not requiring any GPU, just the dummy 0.01)
- m policy actors, all sharing the n GPUs and the 1 replay buffer.

Maybe we should call it "LearnerGroup" or something else. Let me know what you think.